### ISSUE TITLE
Should we set '"types": '[]' for 'tsc init'?

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#39354]
Nobody understands the automatic `@types` inclusion behavior, and it is often a program construction performance foot-gun.

Maybe we should output a field like this in `compilerOptions`?

```json5
// Add globally defined packages that can't be easily imported here.
// Remove entirely to auto-include all '@types' packages.
"types": []
```

**Comment by orta** [comment]
Yes

**Comment by RyanCavanaugh** [comment]
Hmm, also yes

**Comment by danielrentz** [comment]
+1 from me. It was my first step in several TS projects to define "types" fields in every tsconfig.json with explicit type imports to prevent pollution of the code with unneded types from node_modules (e.g., I want @types/node in the grunt scripts, but not in the web scripts, and so on).

**Comment by Bnaya** [comment]
I'm also +1, as setting specific "types" :[] entry (and lib) is the first thing after tsc --init for me
But it might hinder the autocomplete/auto import features of yet-to-be imported packages, so it must be very documented

**Comment by sandersn** [comment]
From discussion in the design meeting: we would like to do this, but only after we've come up with a basic `@types/node` workflow that's as easy as the current one:

1. `import * as fs from "fs"`
2. error "you need to install `@types/node`.
3. `$ npm i @types/node`
4. Done.

Same for jest, when using describe or expect.

**Comment by Bnaya** [comment]
Maybe add new error + autofix for this kind of things?

**Comment by colinhacks** [comment]
> this is really for initializing new projects

Yep I'm aware. My previous post was written with full understanding of this. This is also why I tried to contextualize my points through the lens of a new user following a tutorial (many of which recommend `tsc --init`):

> It's the kind of change that breaks a thousand tutorials in a way where less experienced programmers won't know what went wrong

If the tutorial then has the user install some `@types/*` package without also instruct the user to add the necessary values to `"types"` array, that user will encounter issues.

The current React docs for "Adding TypeScript", for instance, don't mention `"types"` anywhere, because it's generally assumed in educational resources that `"types"` is unset: https://react.dev/learn/typescript#adding-typescript-to-an-existing-react-project

Regarding my anecdotal experience at Bun, we provided a `bun init` command and heavily steered all new users towards it. It generated a tsconfig.json with `types: ["bun-types"]` automatically added. Because "types" was set, it became very easy for users to break their builds by adding declaration files without updating tsconfig.

users would inevitably break their builds by add dependencies & declarations (following some mish-mash of tutorials, probably) and start encountering type resolution errors over time. All the issues I linked to earlier  This happened a lot as you can tell by the volume of issues & reactions to those issues. 

> and it seems like much of the learning curve could be mitigated via error messages

Let's say I forget "node" in my "types" array then try to access `process`. Will TS be able to guess which `@types/*` package needs to be loaded to resolve this global? ("Looks like you're referencing a value from @types/node, but that declaration package isn't being loaded into this workspace. Add `"node"` to the `compilerOptions.types` array.") Seems like a big lift, and probably infeasible. 

But anything short of that wouldn't be specific enough (imo) to mitigate things. Curious what you have in mind here though.

**Comment by HolgerJeromin** [comment]
https://devblogs.microsoft.com/typescript/announcing-typescript-5-9/#minimal-and-updated-tsc---init

Can this issue be closed?

**Comment by RyanCavanaugh** [comment]
Yep!
